### PR TITLE
Feat withdraw function

### LINKED ISSUE #14: [SMART CONTRACT] WRITE THE WITHDRAWAL FUNCTION AND TEST FOR IT
Repository Guidelines

Please star the repository to show your support
Contact the maintainer if you have any questions or need clarification

Task Overview: Bounty Withdrawal Function
Objective
Create a secure function that allows validators and researchers to withdraw their earned bounty payments directly to their wallet addresses, along with comprehensive test coverage.
Function Requirements
Core Functionality
Purpose: Enable authorized users to safely withdraw their earned bounties from completed research projects.
Authorized Users:

Validators - Users who validate research findings
Researchers - Users who conduct and submit research

Key Features to Implement
1. Access Control

Role-Based Permissions: Only validators and researchers should be able to call the withdrawal function
User Verification: Verify that the caller is authorized to withdraw bounties
Address Validation: Ensure withdrawal addresses are valid and belong to the caller

2. Bounty Tracking

Balance Management: Track available bounty amounts for each user
Withdrawal History: Maintain records of all withdrawal transactions
Pending Payments: Handle bounties that may be locked or pending validation

3. Security Measures

Reentrancy Protection: Prevent reentrancy attacks during withdrawal
Amount Validation: Ensure users can only withdraw available balances
Gas Optimization: Efficient function execution to minimize transaction costs
Emergency Controls: Ability to pause withdrawals if security issues arise

4. Payment Processing

Direct Transfer: Send funds directly to user's wallet address
Transaction Verification: Confirm successful transfer before updating balances
Failure Handling: Proper error handling for failed transactions
Event Logging: Emit events for successful withdrawals for transparency

Technical Specifications
Function Parameters

Withdrawal Amount: Amount to withdraw (with validation against available balance)
Recipient Address: Destination wallet address (typically caller's address)
Optional Parameters:

Specific bounty ID (if withdrawing from specific projects)
Payment token type (if multiple tokens supported)

Return Values

Success Status: Boolean indicating successful withdrawal
Transaction Hash: Reference for the withdrawal transaction
Updated Balance: Remaining balance after withdrawal

Error Conditions

Insufficient balance
Unauthorized access (non-validator/researcher)
Invalid recipient address
Contract paused state
Reentrancy attempt
Gas limit exceeded

Comprehensive Testing Strategy
1. Unit Tests
Access Control Tests

Valid Role Access: Test that validators can withdraw successfully
Valid Role Access: Test that researchers can withdraw successfully
Invalid Role Rejection: Test that unauthorized users cannot withdraw
Role Verification: Test role assignment and verification functions

Balance Management Tests

Sufficient Balance: Test withdrawal with adequate balance
Insufficient Balance: Test rejection when balance is too low
Zero Balance: Test behavior with zero balance
Partial Withdrawal: Test withdrawing less than full balance
Full Balance Withdrawal: Test withdrawing entire available balance

Security Tests

Reentrancy Protection: Test that reentrancy attacks are prevented
Double Spending: Test that same bounty cannot be withdrawn twice
Overflow Protection: Test large amount handling
Pause Functionality: Test withdrawal blocking when contract is paused

### PR DESCRIPTION
**Pull request by anonfedora**
# 🔐 Add Bounty Withdrawal System and Emergency Controls

## Overview
This PR implements a comprehensive bounty withdrawal system for validators and approved contributors, along with emergency pause controls for enhanced security and operational flexibility.

Closes #14 

### 1. **Bounty Withdrawal System**
- **New Storage Variables:**
  - `user_bounty_balances: Map<ContractAddress, u256>` - Tracks available bounty balances per user
  - `contract_paused: bool` - Global pause state for emergency control

- **Core Functionality:**
  - `withdraw_bounty(amount: u256, recipient: ContractAddress)` - Allows authorized users to withdraw their earned bounties
  - `add_user_bounty_balance(user: ContractAddress, amount: u256)` - Owner-only function to credit bounty balances

### 2. **Emergency Pause Controls**
- `pause()` - Owner-only function to halt contract operations
- `unpause()` - Owner-only function to resume contract operations
- Withdrawal operations are blocked when contract is paused

### 3. **Authorization System**
- **Dual Authorization Model:**
  - Validators with `VALIDATOR_ROLE` can withdraw bounties
  - Contributors with approved reports can withdraw bounties
- `has_approved_report(user: ContractAddress)` - Internal function to verify contributor approval status

### **Reentrancy Protection**
- Balance updates occur before external calls to prevent reentrancy attacks
- State changes are committed before token transfers

### **Access Control**
- Comprehensive authorization checks for all withdrawal operations
- Owner-only controls for pause/unpause and balance management
- Caller verification prevents unauthorized transfers

### **Input Validation**
- Amount validation (must be > 0)
- Sufficient balance verification
- Recipient address validation (must match caller)

### **Enhanced Payment Processing**
- Improved `process_payment` function with better allowance handling
- Self-contract payments bypass approval requirements
- Robust error messages for debugging

### **Loop Refactoring**
- Replaced `for` loops with more efficient `while` loops for better gas optimization
- Consistent loop patterns across `get_completed_projects_as_array()` and `get_in_progress_projects_as_array()`

### **Event System**
- New `BountyWithdrawn` event with comprehensive logging:
  - User address
  - Withdrawal amount
  - Recipient address
  - Timestamp

### **Comprehensive Test Suite Added:**
1. **Authorization Tests:**
   - `test_withdraw_bounty_unauthorized()` - Verifies unauthorized access prevention
   - `test_withdraw_bounty_invalid_recipient()` - Tests recipient validation

2. **Balance Validation Tests:**
   - `test_withdraw_bounty_insufficient_balance()` - Tests insufficient balance handling
   - `test_withdraw_bounty_zero_balance_validator()` - Zero balance edge case
   - `test_withdraw_bounty_large_amount_insufficient_validator()` - Large amount overflow protection

3. **Edge Case Coverage:**
   - Maximum value testing with `u256::MAX / 2`
   - Zero balance scenarios
   - Invalid recipient scenarios

### **New Public Functions:**
```cairo
fn withdraw_bounty(amount: u256, recipient: ContractAddress) -> (bool, u256)
fn pause()
fn unpause()
fn add_user_bounty_balance(user: ContractAddress, amount: u256)
```

### **Return Values:**
- `withdraw_bounty()` returns `(success: bool, new_balance: u256)` for client-side balance tracking

### **Error Handling:**
- Clear, descriptive error messages for all failure scenarios
- Proper assertion handling with meaningful error codes

### **Gas Optimization:**
- Efficient loop structures
- Minimal storage reads/writes
- Optimized conditional logic

## 🚨 Breaking Changes
- **None** - All changes are additive and maintain backward compatibility

## 🎯 Use Cases

1. **Validator Bounty Claims:**
   - Validators can withdraw earned bounties from successful project validations
   
2. **Contributor Rewards:**
   - Contributors with approved reports can claim their bounty rewards
   
3. **Emergency Response:**
   - Contract owners can pause operations during security incidents
   
4. **Flexible Reward Distribution:**
   - Owners can programmatically distribute bounties to users

## ✅ Testing Recommendations
Screenshot 2025-07-02 at 13 54 09

## 📈 Impact

This enhancement significantly improves the Fortichain ecosystem by:
- Providing secure bounty distribution mechanisms
- Adding emergency controls for operational security
- Maintaining gas efficiency while adding new features
- Ensuring comprehensive access control and validation

---

### CODE REVIEW

File: src/fortichain.cairo [hunk 1]
```diff
@@ -445,9 +456,12 @@ mod Fortichain {
             let token = self.strk_token_address.read();
 
             let erc20_dispatcher = super::IMockUsdcDispatcher { contract_address: token };
-            erc20_dispatcher.approve_user(get_contract_address(), amount);
-            let contract_allowance = erc20_dispatcher.get_allowance(payer, get_contract_address());
-            assert(contract_allowance >= amount, 'INSUFFICIENT_ALLOWANCE');
+            if payer != get_contract_address() {
```
Comment (ONEONUORA) [inline src/fortichain.cairo:459]: The fund is being sent from the escrow

**Review by ONEONUORA — CHANGES_REQUESTED** [review]

**Comment by ONEONUORA** [comment]
Thank you for your contribution @anonfedora
